dedupe/datamodel.py

```diff
@@ -183,30 +178,35 @@ def typify_variables(
         try:
             variable_class = VARIABLE_CLASSES[variable_type]
         except KeyError:
+            valid = ", ".join(VARIABLE_CLASSES)
             raise KeyError(
-                "Field type %s not valid. Valid types include %s"
-                % (definition["type"], ", ".join(VARIABLE_CLASSES))
+                f"Variable type {variable_type} not valid. Valid types include {valid}"
             )
-
         variable_object = variable_class(definition)
-        assert isinstance(variable_object, FieldVariable)
-
-        primary_variables.append(variable_object)
+        assert isinstance(variable_object, Variable)
+        variables.append(variable_object)
 
-        if hasattr(variable_object, "higher_vars"):
-            all_variables.extend(variable_object.higher_vars)
-        else:
-            variable_object = cast(Variable, variable_object)
-            all_variables.append(variable_object)
-
-    if only_custom:
+    no_blocking_variables = all(
+        isinstance(v, (CustomType, InteractionType)) for v in variables
+    )
+    if no_blocking_variables:
         raise ValueError(
-            "At least one of the variable types needs to be a type"
-            "other than 'Custom'. 'Custom' types have no associated"
-            "blocking rules"
+            "At least one of the variable types needs to be a type "
+            "other than 'Custom' or 'Interaction', "
+            "since these types have no associated blocking rules."
         )
 
-    return primary_variables, all_variables
+    return variables
+
+
+def _expand_higher_variables(variables: Iterable[Variable]) -> list[Variable]:
+    result: list[Variable] = []
+    for variable in variables:
+        if hasattr(variable, "higher_vars"):
+            result.extend(variable.higher_vars)
+        else:
+            result.append(variable)
+    return result
```

Comment:
what if we added an __iter__ method to Variable so that we could just do something like.

```
features = []
for variable in variables:
   for feature in variable:
       features.append(feature)
```

Reply:
I think this is a good idea in general, but I would want to name it to differentiate between predicates and features, since Variables supply both. Perhaps `Variable.features`?